### PR TITLE
Add Last batch policy support for remaining image readers

### PR DESCRIPTION
**Pull request by fiona-gladwin**
- Add Last batch policy, pad last batch, stick to shard and shard size support for remaining image readers
- Add changes to COCO, caffe, caffe2, mxnet, tf and cifar10 readers

### CODE REVIEW

**Review by LakshmiKumar23 — APPROVED** [review]

**Comment by kiritigowda** [comment]
@fiona-gladwin can you fix the merge conflicts

**Comment by fiona-gladwin** [comment]
> @fiona-gladwin can you fix the merge conflicts

Fixed merge conflicts

File: rocAL/include/readers/image/caffe_lmdb_record_reader.h [hunk 1]
```diff
@@ -68,6 +68,8 @@ class CaffeLMDBRecordReader : public Reader {
 
     CaffeLMDBRecordReader();
 
+    size_t last_batch_padded_size() override; // The size of the number of samples padded in the last batch
```
Comment (rrawther) [inline rocAL/include/readers/image/caffe_lmdb_record_reader.h:71]: I think we should consider merging the caffe and caffe2 classes. Or derive caffe2 from caffe. I see lots of repeated member variables in both classes

File: rocAL/source/readers/image/caffe_lmdb_record_reader.cpp [hunk 2]
```diff
@@ -38,51 +39,71 @@ CaffeLMDBRecordReader::CaffeLMDBRecordReader()
...
     _last_rec = false;
     _file_count_all_shards = 0;
 }
 
 unsigned CaffeLMDBRecordReader::count_items() {
-    if (_loop)
-        return _file_names.size();
-
-    int ret = ((int)_file_names.size() - _read_counter);
+    int ret = 0;
```
Comment (rrawther) [inline rocAL/source/readers/image/caffe_lmdb_record_reader.cpp:47]: please add new base class and move all common stuff to that class

File: rocAL_pybind/amd/rocal/readers.py [hunk 3]
```diff
@@ -223,7 +223,11 @@ def video(sequence_length, file_list_frame_num=False, file_root="", image_type=t
         "file_list_frame_num": file_list_frame_num}  # VideoMetaDataReader
     b.videoMetaDataReader(Pipeline._current_pipeline._handle,
                           *(kwargs_pybind_reader.values()))
-
+    RocalShardingInfo = b.RocalShardingInfo()
```
Comment (rrawther) [inline rocAL_pybind/amd/rocal/readers.py:226]: please add these defaults into constructor of RocalShardingInfo()

File: rocAL/source/readers/image/caffe_lmdb_record_reader.cpp (hunk 2 above)
Comment (fiona-gladwin) [reply rocAL/source/readers/image/caffe_lmdb_record_reader.cpp]: Done

File: rocAL/include/readers/image/caffe2_lmdb_record_reader.h [hunk 4]
```diff
@@ -68,6 +68,8 @@ class Caffe2LMDBRecordReader : public Reader {
 
     Caffe2LMDBRecordReader();
 
+    size_t last_batch_padded_size() override; // The size of the number of samples padded in the last batch
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply rocAL/include/readers/image/caffe2_lmdb_record_reader.h]: Done, Moved all the common variables and the common functions as protected members to the Reader base class
Introduced new file image_reader.cpp to define the protected member functions of the reader class

File: rocAL_pybind/amd/rocal/readers.py (hunk 3 above)
Comment (fiona-gladwin) [reply rocAL_pybind/amd/rocal/readers.py]: Done

**Comment by fiona-gladwin** [comment]
> please address review comments. Update changelog.md

Addressed review comments and updated Changelog

File: rocAL/include/readers/image/coco_file_source_reader.h [hunk 5]
```diff
@@ -68,7 +68,7 @@ class COCOFileSourceReader : public Reader {
 
    private:
     std::shared_ptr<MetaDataReader> _meta_data_reader = nullptr;
```
Comment (rrawther) [inline rocAL/include/readers/image/coco_file_source_reader.h:70]: please adjust the spacing here. At least 2 spaces from private: to next line

File: rocAL/include/readers/image/image_reader.h [hunk 6]
```diff
@@ -213,5 +213,38 @@ class Reader {
     virtual std::vector<std::string> get_file_paths_from_meta_data_reader() { return {}; }
 
     //! Returns the number of images in the last batch
-    virtual size_t last_batch_padded_size() { return 0; }
+    size_t last_batch_padded_size() { return _last_batch_padded_size; }
+
+   protected:
```
Comment (rrawther) [inline rocAL/include/readers/image/image_reader.h:218]: adjust spacing as before

File: rocAL/include/readers/image/mxnet_recordio_reader.h [hunk 7]
```diff
@@ -67,7 +67,7 @@ class MXNetRecordIOReader : public Reader {
     MXNetRecordIOReader();
 
    private:
-    //! opens the folder containnig the images
+    //! opens the folder containing the images
```
Comment (rrawther) [inline rocAL/include/readers/image/mxnet_recordio_reader.h:70]: adjust spacing

File: rocAL/include/readers/image/mxnet_recordio_reader.h [hunk 8]
```diff
@@ -76,31 +76,20 @@ class MXNetRecordIOReader : public Reader {
...
     unsigned int _last_file_size;
     int64_t _last_seek_pos;
     int64_t _last_data_size;
-    size_t _shard_id = 0;
-    size_t _shard_count = 1;  // equivalent of batch size
-    //!< _batch_count Defines the quantum count of the images to be read. It's usually equal to the user's batch size.
-    /// The loader will repeat images if necessary to be able to have images available in multiples of the load_batch_count,
-    /// for instance if there are 10 images in the dataset and _batch_count is 3, the loader repeats 2 images as if there are 12 images available.
-    size_t _batch_count = 1;
+    size_t _batch_size = 1;
```
Comment (rrawther) [inline rocAL/include/readers/image/mxnet_recordio_reader.h:84]: why is the explanation removed for batch_size

**Review by rrawther — APPROVED** [review]

File: rocAL/include/readers/image/mxnet_recordio_reader.h (hunk 8 above)
Comment (fiona-gladwin) [reply rocAL/include/readers/image/mxnet_recordio_reader.h:84]: Previously, the explanation was for batch count which has now been modified to batch size.
Now the images in the loader will be repeated or dropped based on the Last batch policy and not dependent on the batch count.

**Review by LakshmiKumar23 — APPROVED** [review]

**Pull request merged**